merkle.go

```diff
@@ -0,0 +1,163 @@
+// queried.
+func NewTree() Tree {
+	return &incrementalTree{
+		pendingLeftSiblings: make([]Node, 0),
+		currentLeaf:         0,
+	}
+}
+
+// NewTree creates an empty tree structure that leaves can be added to. While the tree is constructed a single proof is
+// generated that proves membership of all leaves included in leavesToProve. When all leaves have been added the root
+// and proof can be queried.
+func NewProvingTree(leavesToProve []uint64) Tree {
+	return &incrementalTree{
+		pendingLeftSiblings: make([]Node, 0),
+		currentLeaf:         0,
+		leavesToProve:       leavesToProve,
+		proof:               make([]Node, 0),
+	}
+}
+
+func (t *incrementalTree) AddLeaf(leaf Node) {
+	activeNode := leaf
+	for layer := 0; true; layer++ {
+		// If pendingLeftSiblings is shorter than the current layer - extend it.
+		if len(t.pendingLeftSiblings) == layer {
+			t.pendingLeftSiblings = append(t.pendingLeftSiblings, nil)
+		}
+		// If we don't have a node waiting in pendingLeftSiblings, add the active node and break the loop.
+		if t.pendingLeftSiblings[layer] == nil {
+			t.pendingLeftSiblings[layer] = activeNode
+			break
+		}
+		// If the active node should be in the proof - store it.
+		t.addToProofIfNeeded(uint(layer), t.pendingLeftSiblings[layer], activeNode)
+		// Since we found the active node's left sibling in pendingLeftSiblings we can calculate the parent, make it the
+		// active node and move up a layer.
+		activeNode = getParent(t.pendingLeftSiblings[layer], activeNode)
+		// After using the left sibling we clear it from pendingLeftSiblings.
+		t.pendingLeftSiblings[layer] = nil
+	}
+	t.currentLeaf++
+}
+
+func (t *incrementalTree) addToProofIfNeeded(currentLayer uint, leftChild, rightChild Node) {
+	if len(t.leavesToProve) == 0 {
+		// No proof was requested.
+		return
+	}
+	// Calculate the paths to the parent and each child
+	parentPath, leftChildPath, rightChildPath := getPaths(t.currentLeaf, currentLayer)
+	if t.isNodeInProvedPath(parentPath, currentLayer+1) {
+		// We need to be able to calculate the parent.
+		// If the left child isn't in the proved path - we need to include it in the proof.
+		if !t.isNodeInProvedPath(leftChildPath, currentLayer) {
+			t.proof = append(t.proof, leftChild)
+		}
+		// If the right child isn't in the proved path - we need to include it in the proof.
+		if !t.isNodeInProvedPath(rightChildPath, currentLayer) {
+			t.proof = append(t.proof, rightChild)
+		}
+		// It's possible that both children are in the proved path and then we include none of them (and calculate them
+		// instead).
+	}
+}
+
+// getPaths returns the path from the root of the tree to currentLeaf up to the requested layer (parent). It also
+// returns the path to each of its children.
+// Paths are represented as numbers: each binary digit represents a left (0) or right (1) turn.
+func getPaths(currentLeaf uint64, layer uint) (parentPath, leftChildPath, rightChildPath uint64) {
```

Comment:
The name of this method doesn't explain it's purpose well enough. I strongly believe that you should consolidate around one single ID and use it instead of "path", it will improve the code readability. Also, it will be better if this method will only return the parent and the caller will have a way to extract the children (from the parent) using another method